src/Data/TypeRepMap/Internal.hs

```diff
@@ -266,25 +270,32 @@ hoistWithKey f (TypeRepMap as bs ans ks) = TypeRepMap as bs newAns ks
     withTr t = withTypeable t f
 {-# INLINE hoistWithKey #-}
 
--- | The union of two 'TypeRepMap's using a combining function.
+-- | The union of two 'TypeRepMap's using a combining function for conflicting entries
 unionWith :: forall f. (forall x. Typeable x => f x -> f x -> f x) -> TypeRepMap f -> TypeRepMap f -> TypeRepMap f
-unionWith f m1 m2 = fromTriples
-                  $ toTripleList
-                  $ Map.unionWith combine
-                                  (fromTripleList $ toTriples m1)
-                                  (fromTripleList $ toTriples m2)
+unionWith f ma mb = do
+    fromSortedTriples $ mergeMaps (toSortedTriples ma) (toSortedTriples mb)
   where
     f' :: forall x. TypeRep x -> f x -> f x -> f x
     f' tr = withTypeable tr f
 
-    combine :: (Any, Any) -> (Any, Any) -> (Any, Any)
-    combine (av, ak) (bv, _) = (toAny $ f' (fromAny ak) (fromAny av) (fromAny bv), ak)
-
-    fromTripleList :: Ord a => [(a, b, c)] -> Map.Map a (b, c)
-    fromTripleList = Map.fromList . map (\(a, b, c) -> (a, (b, c)))
-
-    toTripleList :: Map.Map a (b, c) -> [(a, b, c)]
-    toTripleList = map (\(a, (b, c)) -> (a, b, c)) . Map.toList
+    combine :: (Fingerprint, Any, Any) -> (Fingerprint, Any, Any) -> (Fingerprint, Any, Any)
+    combine (fp, av, ak) (_, bv, _) = (fp, toAny $ f' (fromAny ak) (fromAny av) (fromAny bv), ak)
+
+    -- Merges two typrepmaps into a sorted, dedup'd list of triples.
+    mergeMaps :: [(Fingerprint, Any, Any)] -> [(Fingerprint, Any, Any)] -> [(Fingerprint, Any, Any)]
+    -- We've addressed all elements from both maps
+    mergeMaps as [] = as
+    mergeMaps [] bs = bs
+    -- Merge
+    mergeMaps (a@(af, _, _) : as) (b@(bf, _, _) : bs)
+      -- Fingerprints are equal, union the elements using our function
+      -- If the incoming maps were de-duped, there shouldn't be any other equivalent
+      -- fingerprints
+      | af == bf = combine a b : mergeMaps as bs
+      -- First fingerprint must not be in the second map or we would have seen it by now
+      -- Add it to the result as-is
+      | af < bf = a : mergeMaps as (b : bs)
+      | otherwise = b : mergeMaps (a:as) bs
```

Comment:
I would replace multiple comparisons with a single call to `compare` function and pattern matching. Otherwise, the merging implementation looks good :+1: